### PR TITLE
Add public ORKSignatureStep

### PR DESCRIPTION
**Pull request by syoung-smallwisdom**
Change the consent signature view controller to a step view controller that is instantiated using an ORKSignatureStep and outputs an ORKSignatureResult. This will allow for customization of the consent process and use of the signature gathering step in other task work flows.

An example implementation can be viewed in `ORKConsentReviewStepViewController`.

Note: This is a simplified version of https://github.com/ResearchKit/ResearchKit/pull/651 that will allow us to use the same functionality that was required by https://github.com/ResearchKit/ResearchKit/pull/733.

### CODE REVIEW

**Comment by oliverschaefer** [comment]
This PR can replace #651

File: ResearchKit/Common/ORKResult.h [hunk 1]
```diff
@@ -1317,4 +1317,25 @@ ORK_CLASS_AVAILABLE
...
+ A signature result is produced by the task view controller when it presents a signature step.
+
+ */
+ORK_CLASS_AVAILABLE
+@interface ORKSignatureResult : ORKResult
+
+/**
+ The signature image generated by this step.
+ */
+@property (nonatomic, copy, nullable) UIImage *signatureImage;
```
Comment (YuanZhu-apple) [inline ResearchKit/Common/ORKResult.h]: No need to copy UIImage

File: ResearchKit/Common/ORKResult.m [hunk 2]
```diff
@@ -2109,3 +2109,54 @@ - (NSString *)descriptionPrefixWithNumberOfPaddingSpaces:(NSUInteger)numberOfPad
...
+
+- (NSUInteger)hash {
+    return [super hash] ^ [self.signatureImage hash];
+}
+
+- (BOOL)isEqual:(id)object {
+    BOOL isParentSame = [super isEqual:object];
+    
+    __typeof(self) castObject = object;
+    return (isParentSame && ORKEqualObjects(self.signatureImage, castObject.signatureImage));
```
Comment (YuanZhu-apple) [inline ResearchKit/Common/ORKResult.m]: Compare the paths as well.

File: ResearchKit/Consent/ORKConsentReviewStepViewController.m [hunk 3]
```diff
@@ -391,6 +396,12 @@ - (void)stepViewControllerResultDidChange:(ORKStepViewController *)stepViewContr
         _signatureLast = (NSString *)lnr.textAnswer;
         [self notifyDelegateOnResultChange];
     }
+    else if ([stepViewController.step.identifier isEqualToString:_SignatureStepIdentifier]) {
```
Comment (YuanZhu-apple) [inline ResearchKit/Consent/ORKConsentReviewStepViewController.m]: Remove newline before `else if`

**Comment by YuanZhu-apple** [comment]
@syoung-smallwisdom Overall it is really good! 
Can you add a test case in ORKTest App?

**Comment by syoung-smallwisdom** [comment]
@YuanZhu-apple Added example to ORKTest.

**Comment by YuanZhu-apple** [comment]
@syoung-smallwisdom In the test, signature is not restored when step is revisited.

**Comment by syoung-smallwisdom** [comment]
@YuanZhu-apple Now restores on back navigation

File: Testing/ORKTest/ORKTest/ORKESerialization.m [hunk 4]
```diff
@@ -1241,6 +1247,11 @@ @implementation ORKESerializer
             PROPERTY(signature, ORKConsentSignature, NSObject, YES, nil, nil),
             PROPERTY(consented, NSNumber, NSObject, YES, nil, nil),
             })),
+   ENTRY(ORKSignatureResult,
+         nil,
+         (@{
+            PROPERTY(signaturePath, UIBezierPath, NSArray, YES, nil, nil),
```
Comment (YuanZhu-apple) [inline Testing/ORKTest/ORKTest/ORKESerialization.m]: Not sure if a `UIBezierPath` can be serialized to JSON string...

Comment (syoung-smallwisdom) [inline Testing/ORKTest/ORKTest/ORKESerialization.m]: @YuanZhu-apple Removed here and added to list of non-serialized properties.

**Comment by YuanZhu-apple** [comment]
r=me.

**Pull request merged**